### ISSUE TITLE
Warning in field calculator when output from expression does not match output field type

### ISSUE THREAD

**Issue by sickel** [qgis/QGIS#56052]
### Feature description

When using the field calculator, an "Output field type" is to be set. The default value is Integer (32 bit). If one forgets to set this correctly when something else is needed, data will be mangled, if this happens, there are (as far as I have found) no simple way to get back to set the correct Output field type. There should be some warning shown when output of the Expression does not match the Output field type. As in the case below, it is clear that the real output will be truncated.
![bilde](https://github.com/qgis/QGIS/assets/3080462/75355da2-ab6a-4f6b-a5d7-e2065aae7fe4)

Maybe something like this quick mockup:
![bilde](https://github.com/qgis/QGIS/assets/3080462/38efb9e2-1d62-4b35-ab52-b188d1637d41)

**Comment by TurboGraphxBeige** [comment]
Yeah I like the idea. The Field Calculator is pretty robust in a way that it will try it's best to accomodate the expression result into the destination field and type. Like you could input an expression like strings '10' + '10' into an Integer field and it would cast the result (1010) into an Integer.

I am just brainstorming here....

Maybe we could use the feature preview to better illustrate what will happen to the result if we were outputting the result of the expression into the chosen field. There could be a yellow warning triangle with a mouse hover tooltip explaining that you have a mismatch between the expression result and the output field type. Like "Warning: Type mismatch between expression output (double) and destination field type (integer)." 

![image](https://github.com/qgis/QGIS/assets/59714546/029bd7f3-5123-460c-a7cb-65292d9e8c8d)

**Comment by blacktesta** [comment]
I have lost my values many times when forgot to set proper field type (for example calculating area with integer field type). Maybe an option to set String as default type in field calculator saves a lot time for those like me because many types can be converted to it. And after you can try to convert it back insted of calculating again.

**Comment by sickel** [comment]
> I have lost my values many times when forgot to set proper field type (for example calculating area with integer field type). Maybe an option to set String as default type in field calculator saves a lot time for those like me because many types can be converted to it. And after you can try to convert it back insted of calculating again.

Good point. Having string as the default data type would ensure one never loses a value.

**Comment by sickel** [comment]
> Yeah I like the idea. The Field Calculator is pretty robust in a way that it will try it's best to accomodate the expression result into the destination field and type. Like you could input an expression like strings '10' + '10' into an Integer field and it would cast the result (1010) into an Integer.
> 
> I am just brainstorming here....
> 
> Maybe we could use the feature preview to better illustrate what will happen to the result if we were outputting the result of the expression into the chosen field. There could be a yellow warning triangle with a mouse hover tooltip explaining that you have a mismatch between the expression result and the output field type. Like "Warning: Type mismatch between expression output (double) and destination field type (integer)."
> 
That would be a good way of showing what is going to happen. My typical case, making a number that should have been stored as a float but setting the output field type to integer i relatively simple. The example you came up with is a bit more difficult, how to handle string concatenations that happens to make a valid number. Maybe a solution would be to use strict typing. So if I want my floating point number to be stored as an integer, or if I am concatenating strings I know will be valid integers and stor them as such, I'll have to use a to_int.

I can also see two simple UI workarounds: As @blacktesta said, if the default output type is String you will hardly ever lose something, Another possibility would be to have the Output Field type blank as default, thereby forcing the user to do a selection.
